// three-forcegraph/dist/three-forcegraph.module.js
import { schemePaired } from 'd3-scale-chromatic';
import tinyColor from 'tinycolor2';
import { Mesh, MeshLambertMaterial, BufferGeometry, BufferAttribute, Matrix4, Vector3, SphereGeometry, CylinderGeometry, Line, LineBasicMaterial, Group } from 'three';
import { forceSimulation, forceLink, forceManyBody, forceCenter } from 'd3-force-3d';
import graph from 'ngraph.graph';
import forcelayout from 'ngraph.forcelayout';
import forcelayout3d from 'ngraph.forcelayout3d';
import Kapsule from 'kapsule';
import accessorFn from 'accessor-fn';

var colorStr2Hex = function colorStr2Hex(str) {
  return isNaN(str) ? parseInt(tinyColor(str).toHex(), 16) : str;
};

// Autoset attribute colorField by colorByAccessor property
// If an object has already a color, don't set it
// Objects can be nodes or links
function autoColorObjects(objects, colorByAccessor, colorField) {
  if (!colorByAccessor || typeof colorField !== 'string') return;

  var colors = schemePaired; // Paired color set from color brewer

  var uncoloredObjects = objects.filter(function (obj) {
    return !obj[colorField];
  });
  var objGroups = {};

  uncoloredObjects.forEach(function (obj) {
    objGroups[colorByAccessor(obj)] = null;
  });
  Object.keys(objGroups).forEach(function (group, idx) {
    objGroups[group] = idx;
  });

  uncoloredObjects.forEach(function (obj) {
    obj[colorField] = colors[objGroups[colorByAccessor(obj)] % colors.length];
  });
}

var classCallCheck = function (instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
};

var inherits = function (subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });
  if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
};

var possibleConstructorReturn = function (self, call) {
  if (!self) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return call && (typeof call === "object" || typeof call === "function") ? call : self;
};

var toConsumableArray = function (arr) {
  if (Array.isArray(arr)) {
    for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) arr2[i] = arr[i];

    return arr2;
  } else {
    return Array.from(arr);
  }
};

var three = window.THREE ? window.THREE // Prefer consumption from global THREE, if exists
: {
  Mesh: Mesh,
  MeshLambertMaterial: MeshLambertMaterial,
  BufferGeometry: BufferGeometry,
  BufferAttribute: BufferAttribute,
  Matrix4: Matrix4,
  Vector3: Vector3,
  SphereGeometry: SphereGeometry,
  CylinderGeometry: CylinderGeometry,
  Line: Line,
  LineBasicMaterial: LineBasicMaterial
};
var ngraph = { graph: graph, forcelayout: forcelayout, forcelayout3d: forcelayout3d };

//

var ForceGraph = Kapsule({

  props: {
    jsonUrl: {},
    graphData: {
      default: {
        nodes: [],
        links: []
      },
      onChange: function onChange(_, state) {
        state.engineRunning = false;
      } // Pause simulation

    },
    numDimensions: {
      default: 3,
      onChange: function onChange(numDim, state) {
        var chargeForce = state.d3ForceLayout.force('charge');
        // Increase repulsion on 3D mode for improved spatial separation
        if (chargeForce) {
          chargeForce.strength(numDim > 2 ? -60 : -30);
        }

        if (numDim < 3) {
          eraseDimension(state.graphData.nodes, 'z');
        }
        if (numDim < 2) {
          eraseDimension(state.graphData.nodes, 'y');
        }

        function eraseDimension(nodes, dim) {
          nodes.forEach(function (d) {
            delete d[dim]; // position
            delete d['v' + dim]; // velocity
          });
        }
      }
    },
    nodeRelSize: { default: 4 }, // volume per val unit
    nodeId: { default: 'id' },
    nodeVal: { default: 'val' },
    nodeResolution: { default: 8 }, // how many slice segments in the sphere's circumference
    nodeColor: { default: 'color' },
    nodeAutoColorBy: {},
    nodeOpacity: { default: 0.75 },
    nodeThreeObject: {},
    linkSource: { default: 'source' },
    linkTarget: { default: 'target' },
    linkColor: { default: 'color' },
    linkAutoColorBy: {},
    linkOpacity: { default: 0.2 },
    linkWidth: {}, // Rounded to nearest decimal. For falsy values use dimensionless line with 1px regardless of distance.
    linkResolution: { default: 6 }, // how many radial segments in each line cylinder's geometry
    linkDirectionalParticles: { default: 0 }, // animate photons travelling in the link direction
    linkDirectionalParticleSpeed: { default: 0.01, triggerUpdate: false }, // in link length ratio per frame
    linkDirectionalParticleWidth: { default: 0.5 },
    linkDirectionalParticleColor: {},
    linkDirectionalParticleResolution: { default: 4 }, // how many slice segments in the particle sphere's circumference
    forceEngine: { default: 'd3' }, // d3 or ngraph
    d3AlphaDecay: { default: 0.0228, triggerUpdate: false, onChange: function onChange(alphaDecay, state) {
        state.d3ForceLayout.alphaDecay(alphaDecay);
      }
    },
    d3AlphaTarget: { default: 0, triggerUpdate: false, onChange: function onChange(alphaTarget, state) {
        state.d3ForceLayout.alphaTarget(alphaTarget);
      }
    },
    d3VelocityDecay: { default: 0.4, triggerUpdate: false, onChange: function onChange(velocityDecay, state) {
        state.d3ForceLayout.velocityDecay(velocityDecay);
      }
    },
    warmupTicks: { default: 0, triggerUpdate: false }, // how many times to tick the force engine at init before starting to render
    cooldownTicks: { default: Infinity, triggerUpdate: false },
    cooldownTime: { default: 15000, triggerUpdate: false }, // ms
    onLoading: { default: function _default() {}, triggerUpdate: false },
    onFinishLoading: { default: function _default() {}, triggerUpdate: false }
  },

  aliases: {
    autoColorBy: 'nodeAutoColorBy'
  },

  methods: {
    // Expose d3 forces for external manipulation
    d3Force: function d3Force(state, forceName, forceFn) {
      if (forceFn === undefined) {
        return state.d3ForceLayout.force(forceName); // Force getter
      }
      state.d3ForceLayout.force(forceName, forceFn); // Force setter
      return this;
    },
    // reset cooldown state
    resetCountdown: function resetCountdown(state) {
      state.cntTicks = 0;
      state.startTickTime = new Date();
      state.engineRunning = true;
      return this;
    },
    tickFrame: function tickFrame(state) {
      var isD3Sim = state.forceEngine !== 'ngraph';

      if (state.engineRunning) {
        layoutTick();
      }
      updatePhotons();

      return this;

      //

      function layoutTick() {
        if (++state.cntTicks > state.cooldownTicks || new Date() - state.startTickTime > state.cooldownTime) {
          state.engineRunning = false; // Stop ticking graph
        } else {
          state.layout[isD3Sim ? 'tick' : 'step'](); // Tick it
        }

        // Update nodes position
        state.graphData.nodes.forEach(function (node) {
          var obj = node.__threeObj;
          if (!obj) return;

          var pos = isD3Sim ? node : state.layout.getNodePosition(node[state.nodeId]);

          obj.position.x = pos.x;
          obj.position.y = pos.y || 0;
          obj.position.z = pos.z || 0;
        });

        // Update links position
        state.graphData.links.forEach(function (link) {
          var line = link.__lineObj;
          if (!line) return;

          var pos = isD3Sim ? link : state.layout.getLinkPosition(state.layout.graph.getLink(link.source, link.target).id);
          var start = pos[isD3Sim ? 'source' : 'from'];
          var end = pos[isD3Sim ? 'target' : 'to'];

          if (line.type === 'Line') {
            // Update line geometry
            var linePos = line.geometry.attributes.position;

            linePos.array[0] = start.x;
            linePos.array[1] = start.y || 0;
            linePos.array[2] = start.z || 0;
            linePos.array[3] = end.x;
            linePos.array[4] = end.y || 0;
            linePos.array[5] = end.z || 0;

            linePos.needsUpdate = true;
            line.geometry.computeBoundingSphere();
          } else {
            // Update cylinder geometry
            var vStart = new three.Vector3(start.x, start.y || 0, start.z || 0);
            var vEnd = new three.Vector3(end.x, end.y || 0, end.z || 0);
            var distance = vStart.distanceTo(vEnd);

            line.position.x = vStart.x;
            line.position.y = vStart.y;
            line.position.z = vStart.z;
            line.lookAt(vEnd);
            line.scale.z = distance;
          }
        });
      }

      function updatePhotons() {
        // update link particle positions
        var particleSpeedAccessor = accessorFn(state.linkDirectionalParticleSpeed);
        state.graphData.links.forEach(function (link) {
          var photons = link.__photonObjs;
          if (!photons || !photons.length) return;

          var pos = isD3Sim ? link : state.layout.getLinkPosition(state.layout.graph.getLink(link.source, link.target).id);
          var start = pos[isD3Sim ? 'source' : 'from'];
          var end = pos[isD3Sim ? 'target' : 'to'];

          var particleSpeed = particleSpeedAccessor(link);

          photons.forEach(function (photon, idx) {
            var photonPosRatio = photon.__progressRatio = ((photon.__progressRatio || idx / photons.length) + particleSpeed) % 1;

            ['x', 'y', 'z'].forEach(function (dim) {
              return photon.position[dim] = start[dim] + (end[dim] - start[dim]) * photonPosRatio || 0;
            });
          });
        });
      }
    }
  },

  stateInit: function stateInit() {
    return {
      d3ForceLayout: forceSimulation().force('link', forceLink()).force('charge', forceManyBody()).force('center', forceCenter()).stop(),
      engineRunning: false
    };
  },

  init: function init(threeObj, state) {
    // Main three object to manipulate
    state.graphScene = threeObj;
  },
  update: function update(state) {
    state.engineRunning = false; // Pause simulation
    state.onLoading();

    if (state.graphData.nodes.length || state.graphData.links.length) {
      console.info('force-graph loading', state.graphData.nodes.length + ' nodes', state.graphData.links.length + ' links');
    }

    if (!state.fetchingJson && state.jsonUrl && !state.graphData.nodes.length && !state.graphData.links.length) {
      // (Re-)load data
      state.fetchingJson = true;

      fetch(state.jsonUrl).then(function (r) {
        return r.json();
      }).then(function (json) {
        state.fetchingJson = false;
        state.graphData = json;
        state._rerender(); // Force re-update
      });
    }

    if (state.nodeAutoColorBy !== null) {
      // Auto add color to uncolored nodes
      autoColorObjects(state.graphData.nodes, accessorFn(state.nodeAutoColorBy), state.nodeColor);
    }
    if (state.linkAutoColorBy !== null) {
      // Auto add color to uncolored links
      autoColorObjects(state.graphData.links, accessorFn(state.linkAutoColorBy), state.linkColor);
    }

    // parse links
    state.graphData.links.forEach(function (link) {
      link.source = link[state.linkSource];
      link.target = link[state.linkTarget];
    });

    // Add WebGL objects
    while (state.graphScene.children.length) {
      state.graphScene.remove(state.graphScene.children[0]);
    } // Clear the place

    var customNodeObjectAccessor = accessorFn(state.nodeThreeObject);
    var valAccessor = accessorFn(state.nodeVal);
    var colorAccessor = accessorFn(state.nodeColor);
    var sphereGeometries = {}; // indexed by node value
    var sphereMaterials = {}; // indexed by color
    state.graphData.nodes.forEach(function (node) {
      var customObj = customNodeObjectAccessor(node);

      var obj = void 0;
      if (customObj) {
        obj = customObj;

        if (state.nodeThreeObject === obj) {
          // clone object if it's a shared object among all nodes
          obj = obj.clone();
        }
      } else {
        // Default object (sphere mesh)
        var val = valAccessor(node) || 1;
        if (!sphereGeometries.hasOwnProperty(val)) {
          sphereGeometries[val] = new three.SphereGeometry(Math.cbrt(val) * state.nodeRelSize, state.nodeResolution, state.nodeResolution);
        }

        var color = colorAccessor(node);
        if (!sphereMaterials.hasOwnProperty(color)) {
          sphereMaterials[color] = new three.MeshLambertMaterial({
            color: colorStr2Hex(color || '#ffffaa'),
            transparent: true,
            opacity: state.nodeOpacity
          });
        }

        obj = new three.Mesh(sphereGeometries[val], sphereMaterials[color]);
      }

      obj.__graphObjType = 'node'; // Add object type
      obj.__data = node; // Attach node data

      state.graphScene.add(node.__threeObj = obj);
    });

    var linkColorAccessor = accessorFn(state.linkColor);
    var linkWidthAccessor = accessorFn(state.linkWidth);
    var linkParticlesAccessor = accessorFn(state.linkDirectionalParticles);
    var linkParticleWidthAccessor = accessorFn(state.linkDirectionalParticleWidth);
    var linkParticleColorAccessor = accessorFn(state.linkDirectionalParticleColor);

    var lineMaterials = {}; // indexed by link color
    var cylinderGeometries = {}; // indexed by link width
    var particleMaterials = {}; // indexed by link color
    var particleGeometries = {}; // indexed by particle width
    state.graphData.links.forEach(function (link) {
      // Add line
      var color = linkColorAccessor(link);
      var linkWidth = Math.ceil(linkWidthAccessor(link) * 10) / 10;

      var useCylinder = !!linkWidth;

      var geometry = void 0;
      if (useCylinder) {
        if (!cylinderGeometries.hasOwnProperty(linkWidth)) {
          var r = linkWidth / 2;
          geometry = new three.CylinderGeometry(r, r, 1, state.linkResolution, 1, false);
          geometry.applyMatrix(new three.Matrix4().makeTranslation(0, 1 / 2, 0));
          geometry.applyMatrix(new three.Matrix4().makeRotationX(Math.PI / 2));
          cylinderGeometries[linkWidth] = geometry;
        }
        geometry = cylinderGeometries[linkWidth];
      } else {
        // Use plain line (constant width)
        geometry = new three.BufferGeometry();
        geometry.addAttribute('position', new three.BufferAttribute(new Float32Array(2 * 3), 3));
      }

      if (!lineMaterials.hasOwnProperty(color)) {
        lineMaterials[color] = new three.MeshLambertMaterial({
          color: colorStr2Hex(color || '#f0f0f0'),
          transparent: true,
          opacity: state.linkOpacity
        });
      }
      var lineMaterial = lineMaterials[color];

      var line = new three[useCylinder ? 'Mesh' : 'Line'](geometry, lineMaterial);

      line.renderOrder = 10; // Prevent visual glitches of dark lines on top of nodes by rendering them last

      line.__graphObjType = 'link'; // Add object type
      line.__data = link; // Attach link data

      state.graphScene.add(link.__lineObj = line);

      // Add photon particles
      var numPhotons = Math.round(Math.abs(linkParticlesAccessor(link)));
      var photonR = Math.ceil(linkParticleWidthAccessor(link) * 10) / 10 / 2;
      var photonColor = linkParticleColorAccessor(link) || color || '#f0f0f0';

      if (!particleGeometries.hasOwnProperty(photonR)) {
        particleGeometries[photonR] = new three.SphereGeometry(photonR, state.linkDirectionalParticleResolution, state.linkDirectionalParticleResolution);
      }
      var particleGeometry = particleGeometries[photonR];

      if (!particleMaterials.hasOwnProperty(color)) {
        particleMaterials[color] = new three.MeshLambertMaterial({
          color: colorStr2Hex(photonColor),
          transparent: true,
          opacity: state.linkOpacity * 3
        });
      }
      var particleMaterial = particleMaterials[color];

      var photons = [].concat(toConsumableArray(Array(numPhotons))).map(function () {
        return new three.Mesh(particleGeometry, particleMaterial);
      });
      photons.forEach(function (photon) {
        return state.graphScene.add(photon);
      });
      link.__photonObjs = photons;
    });

    // Feed data to force-directed layout
    var isD3Sim = state.forceEngine !== 'ngraph';
    var layout = void 0;
    if (isD3Sim) {
      // D3-force
      (layout = state.d3ForceLayout).stop().alpha(1) // re-heat the simulation
      .numDimensions(state.numDimensions).nodes(state.graphData.nodes).force('link').id(function (d) {
        return d[state.nodeId];
      }).links(state.graphData.links);
    } else {
      // ngraph
      var _graph = ngraph.graph();
      state.graphData.nodes.forEach(function (node) {
        _graph.addNode(node[state.nodeId]);
      });
      state.graphData.links.forEach(function (link) {
        _graph.addLink(link.source, link.target);
      });
      layout = ngraph['forcelayout' + (state.numDimensions === 2 ? '' : '3d')](_graph);
      layout.graph = _graph; // Attach graph reference to layout
    }

    for (var i = 0; i < state.warmupTicks; i++) {
      layout[isD3Sim ? 'tick' : 'step']();
    } // Initial ticks before starting to render

    state.layout = layout;
    this.resetCountdown();
    state.onFinishLoading();
  }
});

function fromKapsule (kapsule) {
  var baseClass = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : Object;
  var initKapsuleWithSelf = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;

  var FromKapsule = function (_baseClass) {
    inherits(FromKapsule, _baseClass);

    function FromKapsule() {
      var _ref;

      classCallCheck(this, FromKapsule);

      for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      var _this = possibleConstructorReturn(this, (_ref = FromKapsule.__proto__ || Object.getPrototypeOf(FromKapsule)).call.apply(_ref, [this].concat(args)));

      _this.__kapsuleInstance = kapsule().apply(undefined, [].concat(toConsumableArray(initKapsuleWithSelf ? [_this] : []), args));
      return _this;
    }

    return FromKapsule;
  }(baseClass);

  // attach kapsule props/methods to class prototype


  Object.keys(kapsule()).forEach(function (m) {
    return FromKapsule.prototype[m] = function () {
      var _kapsuleInstance;

      var returnVal = (_kapsuleInstance = this.__kapsuleInstance)[m].apply(_kapsuleInstance, arguments);

      return returnVal === this.__kapsuleInstance ? this // chain based on this class, not the kapsule obj
      : returnVal;
    };
  });

  return FromKapsule;
}

var threeForcegraph = fromKapsule(ForceGraph, Group, true);

export default threeForcegraph;
